### ISSUE TITLE
Log an object that contains messages with a new method add_loggable_object( )

### ISSUE THREAD

**Issue by AlexandreHT** [ABAP-Logger/ABAP-Logger#67]
### Use case
The method validate( ) of a model class performs checks and returns an object with all the errors found. (see https://www.martinfowler.com/articles/replaceThrowWithNotification.html for the pattern description).

I may display the messages to the users but I may also log them. 
the model class could use a zcl_logger instance but the model does not know whether we are in a validating user input (with one call per PAI/PBO cycle) or processing some code which need to be logged for good.

### Current
As of today, I get the messages, loop on them and log each one of them.

### Proposed solution
  * create a new interface zif_loggable_object with one method get_message_table
  * create a new method zif_logger~add_loggable_object with one parameter object type ref to zif_loggable_object
  * the definition of the class that implements zif_loggable_object would left to each one of us

### Target use
Validation( ) method returns an object that implements zif_loggable_object. The upper-level code may log these message in one instruction such as 
```ABAP
data(notification) = model->validate( ).
if notification->has_error( ) = abap_true.
  logger->add_loggable_object( notification ).
endif.
```

Please do not hesitate to provide feedback on this proposition, the naming or anything, 
Thanks,

**Comment by andkopp** [comment]
Interesting approach. Can you extend your example on how the validate() method would add loggable messages to the notification (zif_loggable_object)? I assume that the implementation could be a repetition of code that is already implemented in zcl_logger like add(), e()...

**Comment by AlexandreHT** [comment]
You make a point. I was about to write the following enhancement proposition: I think we should extract this code in a functional class (eg ``zcl_abap_logger_tools``). 
I see the following advantages:
  * it would simplify zcl_abap_logger
  * it would allow these functions to be used elsewhere
  * it would match the following Clean ABAP recommendation [Don't mix stateful and stateless in the same class](https://github.com/SAP/styleguides/blob/master/clean-abap/CleanABAP.md#dont-mix-stateful-and-stateless-in-the-same-class)

To my opinion, the RTTS mapping code in the add method is the most valuable part of ABAP-Logger. It could be a co-product of the main zif_logger capability (as a matter of fact, I have multiple logging methods in my company and it would be nice to use this tool class).

NB: I think merging #49 into master is a prerequisite to this.

**Comment by andkopp** [comment]
I agree. RTTS mapping is the most valuable part. And I also agree on avoiding mixing of stateful and stateless code.
So `zcl_logger` would continue to implement `zif_logger` and delegate calls on `add()` to `zcl_logger_tools`, correct? Probably not an easy task to design the interface between separate those classes.

**Comment by AlexandreHT** [comment]
I was thinking about the following methods:
- map_sy_to_msg
    -  returning a structure with ty/id/no/v1/v2/v3/v4
- is_struct_name_mapbl_to_msg
    - importing a RTTS absolute name (type abap_abstypename)
    - returning a boolean
- is_struct_mappable_to_msg
    - importing a structure
    - returning a boolean
- map_structure_to_msg
    - importing a structure
    - returning  a structure with ty/id/no/v1/v2/v3/v4
- map_structure_to_string_table : replaces add_structure
    - importing a (deep) structure
    - returning  a table of strings
- drill_down_into_exception
    - importing exception type ref to cx_root
    - returning a table of exceptions
- map
  - importing any
  - exporting a structure with ty/id/no/v1/v2/v3/v4
  - exporting a table of strings
  - exporting a table of exceptions

Maybe zcl_abap_logger_data_mapper would be more suitable as a class name.

**Comment by andkopp** [comment]
This makes it much clearer. I like your approach.

**Comment by mbtools** [comment]
Closed by #90
